odl/phantom/phantom_utils.py

```diff
@@ -0,0 +1,254 @@
+﻿# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Utilities for phantoms."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+
+
+__all__ = ('ellipse_phantom', 'cylinders_from_ellipses')
+
+
+def cylinders_from_ellipses(ellipses2d):
+    """Create 3d cylinders from ellipses."""
+    ellipses2d = np.asarray(ellipses2d)
+    ellipses3d = np.zeros((ellipses2d.shape[0], 10))
+    ellipses3d[:, [0, 1, 2, 4, 5, 7]] = ellipses2d
+    ellipses3d[:, 3] = 100000.0
+
+    return ellipses3d
+
+
+def ellipse_phantom_2d(space, ellipses):
+    """Create an ellipse phantom in 2d space.
+
+    Parameters
+    ----------
+    space : `DiscreteLp`
+        The space the phantom should be generated in.
+    ellipses : list of lists
+        Each row should contain:
+        'value', 'axis_1', 'axis_2', 'center_x', 'center_y', 'rotation'
+        The ellipses should be contained the he rectangle [-1, -1] x [1, 1].
+
+    Returns
+    -------
+    phantom : `DiscreteLpVector`
+        The phantom
+    """
+
+    # Blank image
+    p = np.zeros(space.size)
+
+    # Create the pixel grid
+    points = space.points()
+    minp = space.grid.min()
+    maxp = space.grid.max()
+
+    # move points to [0, 1]
+    points = (points - minp) / (maxp - minp)
+
+    # move to [-1, 1]
+    points = points * 2 - 1
+
+    for ellip in ellipses:
+        intensity = ellip[0]
+        a_squared = ellip[1] ** 2
+        b_squared = ellip[2] ** 2
+        x0 = ellip[3]
+        y0 = ellip[4]
+        phi = ellip[5] * np.pi / 180.0
+
+        # Create the offset x and y values for the grid
+        offset_points = points - [x0, y0]
+
+        cos_p = np.cos(phi)
+        sin_p = np.sin(phi)
+
+        # Find the pixels within the ellipse
+        scales = [1 / a_squared, 1 / b_squared]
+        mat = [[cos_p, sin_p],
+               [-sin_p, cos_p]]
+        radius = np.dot(scales, np.dot(mat, offset_points.T) ** 2)
+        inside = radius <= 1
+
+        # Add the ellipse intensity to those pixels
+        p[inside] += intensity
+
+    return space.element(p)
```

Comment:
Ok, so it seems the code was pretty trivial to port. With my computer I got 3.25s with the old code, and 0.17s with the new code on your example.